### PR TITLE
[3626] Reactify Request Publish dialog

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/3626

### CODE REVIEW

File: ui/app/package.json [hunk 1]
```diff
@@ -17,11 +19,13 @@
     "@types/video.js": "^7.2.10",
     "@videojs/http-streaming": "^1.10.3",
     "clsx": "^1.0.4",
+    "date-fns": "^2.0.0-beta.5",
```
Comment (rart) [inline ui/app/package.json:25]: Are both date-fns needed? date-fns and @date-io/date-fns

File: ui/app/src/modules/Content/Approve/ApproveDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+import moment from "moment";
+
+const messages = defineMessages({
+  title: {
+    id: 'approveDialog.title',
+    defaultMessage: 'Approve for Publish'
+  },
+  subtitle: {
+    id: 'approveDialog.subtitle',
+    defaultMessage: 'Selected files will go live upon submission. Hard dependencies are automatically submitted with the' +
```
Comment (rart) [inline ui/app/src/modules/Content/Approve/ApproveDialog.tsx]: Space missing between the and main

File: ui/app/src/modules/Content/Dependencies/DependencySelection.tsx [hunk 3]
```diff
@@ -417,67 +343,97 @@ function SelectionList(props: SelectionListProps) {
...
+          {
+            items.map((item: Item) => {
+              const labelId = `checkbox-list-label-${item.uri}`;
+
+              return (
+                <ListItem
+                  key={item.uri}
+                  role={undefined}
+                  // dense
+                  {...onItemClicked ? {
```
Comment (rart) [inline ui/app/src/modules/Content/Dependencies/DependencySelection.tsx]: This expression looks very suspicious. I don’t think you can spread null.

Comment (rart) [reply ui/app/src/modules/Content/Dependencies/DependencySelection.tsx]: Put some parenthesis around the ternary expression to make this clearer.

File: ui/app/src/modules/Content/Dependencies/DependencySelection.tsx [hunk 4]
```diff
@@ -417,67 +343,97 @@ function SelectionList(props: SelectionListProps) {
...
-              ))
+              uris.map((uri: string) => {
+                const labelId = `checkbox-list-label-${uri}`;
+
+                return (
+                  <ListItem
+                    key={uri}
+                    role={undefined}
+                    dense
+                    {...onItemClicked ? {
```
Comment (rart) [inline ui/app/src/modules/Content/Dependencies/DependencySelection.tsx]: 😱 same as above

File: ui/app/src/modules/Content/Submit/DateTimePicker.tsx [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+]
+
+const DateTimePicker = withStyles(dateTimePickerStyles)((props: DateTimePickerProps) => {
+  const {
+    classes,
+    onChange,
+    onChangeDate,
+    onChangeTime,
+    onChangeTimezone,
+    initialDate,
```
Comment (rart) [inline ui/app/src/modules/Content/Submit/DateTimePicker.tsx]: initailDate= moment()

the on the useState you can simply use initialDate. Same for other props like time zone and controls. Provide default / fallback values.

For example, with controls, you would no longer need the showall var, nor check if controls is null; just check if it includes the control you wish to check for.

File: ui/app/src/modules/Content/Submit/DateTimePicker.tsx [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+          inputProps={{
+            className: classes.select
+          }}
+          onChange={handleSelectChange()}
+          IconComponent={ PublicIcon }
+          classes={{
+            icon: classes.selectIcon
+          }}
+        >
+          { timezones &&
```
Comment (rart) [inline ui/app/src/modules/Content/Submit/DateTimePicker.tsx]: Format

File: ui/app/src/modules/Content/Submit/RequestPublishDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+  deps: any,
+  showDepsButton: boolean,
+  selectAllDeps: Function,
+  selectAllSoft: Function,
+  onClickShowAllDeps?: any,
+  showEmailCheckbox?: boolean
+  classes?: any;
+}
+
+export const PublishDialogUI = withStyles(dialogStyles)((props: PublishDialogUIProps) => {
```
Comment (rart) [inline ui/app/src/modules/Content/Submit/RequestPublishDialog.tsx]: Move to separate file. (IntelliJ does this for you brilliantly: use Refactor > Move)

File: ui/app/src/services/content.ts [hunk 8]
```diff
@@ -27,10 +26,17 @@ export function getContent(site: string, path: string): Observable<string> {
 }
 
 export function getDOM(site: string, path: string): Observable<XMLDocument> {
-  return getContent(site, path).pipe(map(fromString));
```
Comment (rart) [inline ui/app/src/services/content.ts]: Looks like you accepted yours on the merge. You overwrote recent changes. Rollback.

File: ui/app/src/services/publishing.ts [hunk 9]
```diff
@@ -36,9 +36,25 @@ export function fetchEnvironments(siteId: string) {
   return get(`/studio/api/1/services/api/1/deployment/get-available-publishing-channels.json?site_id=${siteId}`)
 }
 
+export function submitToGoLive(siteId: string, user:string, data) {
+  return post(`/studio/api/1/services/api/1/workflow/submit-to-go-live.json?site=${siteId}&user=${user}`, data,
```
Comment (rart) [inline ui/app/src/services/publishing.ts]: Format: all params on its own line

```
post(
  p1
  p2
  p3
)
```

Comment (rart) [reply ui/app/src/services/publishing.ts]: same for similar instances on thispr

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/package.json (hunk 1 above)
Comment (jvega190) [reply ui/app/package.json:25]: I tested and it seems like `date-fns` is used by `@date-io/date-fns`.

File: static-assets/components/cstudio-common/common-api.js [hunk 10]
```diff
@@ -830,14 +830,19 @@ var nodeOpen = false,
...
-                        CSA.Operations.translateContent(formsLangBundle, ".cstudio-dialogue");
-                        this.loadItems(items, dialogue);
-                    }
-                }, true, '800px', approveType);
+              const container = ($('<div class="approve-dialog-container"/>').appendTo('body'))[0];
+
+              const onClose = () => {
+                CrafterCMSNext.ReactDOM.unmountComponentAtNode(container);
+              };
+
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js]: render function now provides the unmount function
```js
let unmount;
CrafterNext.render({
...
onClose: () => unmount()
...
}).then(done => unmount = done.unmount)
```

File: static-assets/components/cstudio-common/common-api.js [hunk 11]
```diff
@@ -859,14 +864,21 @@ var nodeOpen = false,
...
             submitContent: function(site, items) {
-                CSA.Operations._showDialogueView({
-                    fn: CSA.Service.getRequestPublishView,
-                    controller: 'viewcontroller-requestpublish',
-                    callback: function(dialogue) {
-                        CSA.Operations.translateContent(formsLangBundle, ".cstudio-dialogue");
-                        this.loadItems(items, dialogue);
-                    }
-                }, true, '800px');
+                // CSA.Operations._showDialogueView({
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js]: x

File: static-assets/components/cstudio-common/common-api.js [hunk 12]
```diff
@@ -859,14 +864,21 @@ var nodeOpen = false,
...
+                //     fn: CSA.Service.getRequestPublishView,
+                //     controller: 'viewcontroller-requestpublish',
+                //     callback: function(dialogue) {
+                //         CSA.Operations.translateContent(formsLangBundle, ".cstudio-dialogue");
+                //         this.loadItems(items, dialogue);
+                //     }
+                // }, true, '800px');
+              var container = ($('<div class="request-publish-container"></div>').appendTo('body'))[0];
+              CrafterCMSNext.render(container, 'RequestPublishDialog',
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js]: onclose

File: ui/app/package.json [hunk 13]
```diff
@@ -4,7 +4,9 @@
   "private": true,
   "dependencies": {
     "@material-ui/core": "^4.7.2",
+    "@date-io/date-fns": "^1.3.11",
```
Comment (rart) [inline ui/app/package.json:6]: check for updates on all new deps

File: static-assets/components/cstudio-common/common-api.js [hunk 14]
```diff
@@ -859,14 +874,60 @@ var nodeOpen = false,
...
+                  onClose: (response) => {
+                    if(response) {
+                      _self.reloadItems(items, response);
                     }
-                }, true, '800px');
+
+                    unmount();
+                  },
+                  items: items,
+                  siteId: site,
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js]: site and user may now be retrieved from the react context store

File: ui/app/src/components/DateTimePicker.tsx [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+      case 'scheduledTime':
+        updatedDateTime.hours(date.getHours());
+        updatedDateTime.minutes(date.getMinutes());
+        break;
+    }
+
+    setSelectedDateTime(updatedDateTime);
+    onChange && onChange(updatedDateTime);
+
+    onChangeDate && onChangeDate(updatedDateTime.format('YYYY-MM-DD'));
```
Comment (rart) [inline ui/app/src/components/DateTimePicker.tsx]: Date and time formats should be props. Maybe compute datetime format from the two.

Also:
`onChange?.(updatedDateTime...)`

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/package.json (hunk 13 above)
Comment (jvega190) [reply ui/app/package.json:6]: @rart none of the new deps have updates, but I saw some others outdated, should I update?

**Review by rart — APPROVED** [review]

**Review by avasquez614 — APPROVED** [review]

**Pull request merged**